website/blog/2025/05-05-arbiter-type-class/index.md

```diff
@@ -0,0 +1,280 @@
+) // the two function arguments are the same as above.
+arbiter2.clients :<>= clients3
+```
+
+## Introducing a Type Class
+
+To clean this up even more, we can introduce a type class that captures the "arbitrable" pattern:
+
+```scala
+trait Arbitrable[A] {
+  def request(a: A): Bool
+  def grant(a: A, value: Bool): Unit
+}
+```
+
+Effectively, we have taken the two arguments to the arbiter and turned them into methods on the type class.
+This looks similar to the proposed object-oriented version of `Arbitrable` above, but note how it is parameterized by the type of the client and accepts the client as an argument.
+
+We can then provide instances of this type class for specific types. For example, for `ArbiterClient` and `Decoupled`:
+
+```scala
+class ArbiterClientArbitrable extends Arbitrable[ArbiterClient] {
+  def request(a: ArbiterClient) = a.request
+  def grant(a: ArbiterClient, value: Bool) = a.grant := value
+}
+
+class DecoupledArbitrable[T <: Data] extends Arbitrable[DecoupledIO[T]] {
+  def request(a: DecoupledIO[T]) = a.valid
+  def grant(a: DecoupledIO[T], value: Bool) = a.ready := value
+}
+```
+
+Then, we can refactor the arbiter to use the type class:
+
+```scala
+class GenericPriorityArbiter[A <: Data](nClients: Int, clientType: A, arbitrable: Arbitrable[A]) extends Module {
+  val clients = IO(Vec(nClients, Flipped(clientType)))
+
+  val requests = clients.map(arbitrable.request(_))
+  val granted = PriorityEncoderOH(requests)
+
+  for ((client, grant) <- clients.zip(granted)) {
+    arbitrable.grant(client, grant)
+  }
+}
+```
+
+This makes instantiating the arbiter a hair nicer:
+
+```scala
+val clients1 = IO(Vec(4, Flipped(new ArbiterClient)))
+val arbiter1 = Module(new GenericPriorityArbiter(4, new ArbiterClient, new ArbiterClientArbitrable))
+arbiter1.clients :<>= clients1
+
+val clients2 = IO(Vec(4, Flipped(Decoupled(UInt(8.W)))))
+val arbiter2 = Module(new GenericPriorityArbiter(4, Decoupled(UInt(8.W)), new DecoupledArbitrable[UInt]))
+arbiter2.clients :<>= clients2
+```
+
+At least we aren't repeating logic anymore, instead we get to just refer to the type class instance.
+
+However, we can do even better.
+
+## Implicit Type Class Instances
+
+Scala has a powerful feature called **implicit resolution**.
+This allows us to avoid passing around the type class instance explicitly.
+Instead, we can define the type class instance as an implicit value and the compiler will automatically find it for us.
+
+Let us rewrite our typeclass instances as implicit values:
+
+```scala
+implicit val arbiterClientArbitrable: Arbitrable[ArbiterClient] =
+  new Arbitrable[ArbiterClient] {
+    def request(a: ArbiterClient) = a.request
+    def grant(a: ArbiterClient, value: Bool) = a.grant := value
+  }
+
+// In chisel3.util, the type is DecoupledIO while we construct instances of it with Decoupled.
+// Note that this is a def because DecoupledIO itself takes a type parameter.
```

Comment:
```suggestion
// Note that this is a def because DecoupledIO itself takes a type parameter,
// so we can't reuse the same one for every call-site.
```